svelte: Redesign file header

fkling opened the pull request:
Closes #61652
Closes #61630
Closes #61631

This PR implements the redesigned file header. Specifically it:

- Updates the font and spacing of the file path
- Adds the file icon to the file name
- Updates the styles of the file action buttons/links
- Adds an overflow menu for less important actions
- Adds an accessible "view mode" switcher
- Adds file meta information (lines and size)

NOTE: Since blame view is not supported yet I've removed that option for now, but it's easy to add back.

NOTE 2: As pointed out in #61652, I intentionally excluded some parts in this first iteration.

| Formatted with open menu | Code |
|--------|--------|
| ![2024-04-09_13-53](https://github.com/sourcegraph/sourcegraph/assets/179026/87b96a98-53fe-4a04-8b1d-3a4a3a89c033) | ![2024-04-09_13-49_1](https://github.com/sourcegraph/sourcegraph/assets/179026/bbbfd8fd-f815-4530-84cd-29f982a7e6dc) | 
| Note: The screenshot shows the menu entry as "Raw download" but I've changed it to "View raw" ||

Unlike the React version, "View raw" opens the file in a new tab instead of directly downloading it. That's something we can iterate on in the future.

I've also adjusted removed the icons from the headers on the repo home page and the directory page, to be more aligned with the file page.

## Test plan

- Tested file actions (open on codehost, permalink, view raw, line wrapping)
- Tested view mode changes and switcher UI for "formatted" files (e.g. markdown) files, and code files.
- Visual inspection of repo home and directory pages.
- New playwright tests for file page

fkling commented on client/web-sveltekit/src/lib/repo/Actions.svelte:
```diff
@@ -0,0 +1,59 @@
+            align-self: center;
+        }
+
+        @container fileheader (width < 50cqw) {
+            :global([data-action-label]) {
+                display: none;
+            }
+        }
```
This is a very simple way to hide the button/link labels when the container has a certain size. It's not ideal since it doesn't take the length of the file path into account, but again, it's a simple CSS based solution.

fkling commented on client/web-sveltekit/src/lib/repo/FileHeader.svelte:
```diff
@@ -13,39 +13,40 @@
+            <span class:last={index === breadcrumbs.length - 1}>
                 {#if index > 0}
                     /
                 {/if}
-                <span class:last={index === breadcrumbs.length - 1}>
-                    {#if path}
-                        <a href={path}>{name}</a>
-                    {:else}
-                        {name}
-                    {/if}
-                </span>
-            {/each}
-        </span>
+                {#if path}
+                    <a href={path}>{name}</a>
+                {:else}
+                    <slot name="icon" />
+                    {name}
+                {/if}
+            </span>
+        {/each}
```
I should add another comment here: I restructured this so that a path like

```
/very/long/path/file.ts
```

is wrapped like this

```
/very/long/path
/file.ts
```

I.e. the slash before the segment will always be wrapped *with* the segment.

camdencheek commented on client/web-sveltekit/src/lib/wildcard/menu/MenuButton.svelte:
```diff
@@ -0,0 +1,14 @@
+
+    type $$Props = HTMLButtonAttributes
+
+    const { item } = getContext()
```
I think this is unrelated to the PR, but why do we communicate this via context?

fkling replied:
`DropdownMenu.svelte` uses melt UI to create some builders for menu items, submenus, etc and they are specific to the menu that's created. Context is the simplest way to pass those builders to any descendants.

camdencheek replied:
Ah, got it. Thanks!

fkling replied:
Slot props would be a bit round about. The callsite doesn't need to worry about the inner workings/relation between `DropdownMenu` and `MenuButton`/`MenuLink`.

camdencheek commented on client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/-/blob/[...path]/+page.svelte:
```diff
@@ -52,29 +61,86 @@
     }
+
+    function changeViewMode({ detail: viewMode }: { detail: ViewMode }) {
+        console.log('changeViewMode', viewMode)
```
intentional?

camdencheek commented on client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/-/blob/[...path]/+page.svelte:
```diff
@@ -52,29 +61,86 @@
+                } else {
+                    url.deleteSearchParameter('view')
+                }
+                goto(url.toString(), { replaceState: true, keepFocus: true })
```
One thing I noticed: clicking `Code` triggers a navigation, which means we jump into full-page mode. That was a little surprising to me, especially since the button moved.

fkling replied:
Yep, I noticed that too. Might have to make an exception when changing the view.

fkling replied:
I don't know why but that seems to only happen when I click on the view mode buttons after a fresh page load. When I navigate to/from another file it doesn't happen (and it shouldn't according to the code). So that "to the top scrolling" behavior seems to coming from somewhere else. We should just remove full page scrolling.

camdencheek commented on client/web-sveltekit/src/routes/[...repo=reporev]/(validrev)/(code)/-/blob/[...path]/+page.svelte:
```diff
@@ -52,29 +61,86 @@
+                    {/if}
+                    <Permalink {commitID} />
+                    <svelte:fragment slot="menu">
+                        <MenuLink href="{repoURL}/-/raw/{filePath}" target="_blank">
+                            <Icon svgPath={mdiFileEyeOutline} inline /> View raw
+                        </MenuLink>
```
Unrelated to this PR: I just noticed we already have a raw file endpoint. RE: our previous conversation about caching, can we just use that for fetching file contents?

fkling replied:
I don't see why not.

fkling commented on client/web-sveltekit/src/lib/dom.ts:
```diff
@@ -16,6 +16,70 @@ import * as uuid from 'uuid'
+    while (true) {
+        while (sibling) {
+            if (sibling.matches(selector)) {
+                return sibling as T
```
Yes, but I think it's "normal" with DOM APIs. E.g. if you use something like `querySelector` to find a specific element, or access `event.target` in an event handler, the results are often cast to a specific element type.
Likewise here we assume that the caller knows that the selector matches only a specific type of element.
It's not great, but I guess that's what we have :(

camdencheek replied:
Makes sense 👍 

Wondering then if we should move the casting to the caller though to force them to be explicit that they know the type rather than assuming they know that passing a type parameter means they're asserting that their selector can only return elements of that type.

fkling commented on client/web-sveltekit/src/lib/repo/Permalink.svelte:
```diff
@@ -12,7 +16,15 @@
 {#if href}
-    <Tooltip tooltip="Permalink (with full Git commit SHA)">
-        <a {href}><Icon svgPath={mdiLink} inline /></a>
+    <Tooltip tooltip="Permalink (with full git commit SHA)">
+        <a {href}><Icon svgPath={mdiLink} inline /> <span data-action-label>Permalink</span></a>
```
Yes. Happy to change this in another PR but I'd also like to put this up for discussion. I really like links ;)

fkling commented on client/web-sveltekit/src/lib/repo/Actions.svelte:
```diff
@@ -0,0 +1,59 @@
+            align-self: center;
+        }
+
+        @container fileheader (width < 50cqw) {
+            :global([data-action-label]) {
+                display: none;
+            }
+        }
```
You are right, I think I need to change what the container is supposed to be. I tried various things so maybe things got mixed up a bit.

fkling replied:
But to clarify, it will never work perfectly, i.e. there is not way to determine with CSS alone whether or not there is enough space for the labels. But at least we can make it relative to the size of the file header.
I had tried a JS based approach but it was kind of broken, so I thought before I spent more time on this specifically I rather get some approximation working and we'll revisit this another time.

fkling replied:
We can do that with container queries. What I meant is that ideally the labels would be hidden as soon as the path starts to wrap. That's not something we can do in CSS alone.

camdencheek replied:
Oooh, gotcha. Thanks for clarifying 👍

fkling replied:
As a quick update: I've misunderstood some aspects of container queries and it doesn't work like I expected it to. The main issue is that by creating a "containment context", the container looses the ability to compute its own size by its contents, which is useful in a flex box context. But I'll experiment a bit with different settings.

fkling commented on client/web-sveltekit/src/lib/wildcard/menu/DropdownMenu.svelte:
```diff
@@ -16,15 +16,15 @@
-    interface $$Props extends HTMLButtonAttributes {
-        open: Writable<boolean>
+    type $$Props = {
+        open?: Writable<boolean>
         triggerButtonClass: string
-    }
+    } & HTMLButtonAttributes
```
Moving from `interface` to `type`? In my editor setup I get warnings that `$$Props` cannot be overwritten or something when it's an interface. Honestly I don't know why. This doesn't trigger the error but also works as expected 🤷🏻‍♂️

camdencheek replied:
Oh cool. I also get that error. Will start using `type` 😄

fkling commented:
OK, I completely changed how we determine the space available for actions. It didn't seem there was an easy way to make this work with container queries, even just approximating it, due to how creating containment contexts impact the size calculation of the container.

Instead I want back to a JS based approach. Due to how different styles can impact the size and layout of an element it seems difficult to find a universally applicable approach. So with the API added here it's the callers responsibility to define/provide styles that allow the utility to measure the element's size. I hope the comments and the example explain the approach well enough.

Additionally I removed the `Actions.svelte` class and moved the actions logic into the file header instead. The reason being that there didn't seem to be a way to prevent leaking spacing settings (e.g. gaps, padding) between the two elements (either the actions component needs to know about the padding the file header applies or the file header needs to know which gap the actions component uses). To me that was an indication that this should really be a single component.

taiyab commented:
Beautiful work @fkling!
